### PR TITLE
Сapture Python output reliably

### LINKED ISSUE #663: ProcessHelper occasionally misses data and GetSearchPathsAsync fails
This is because next event can come while previous is still being processed and hence sometimes Dispose() is called before all events were handled.

https://github.com/Microsoft/python-language-server/pull/653 made this much more visible.

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Fixes #663 

The issue was that events from the process come in multiple threads and next event can come while previous is still being handled (and get handled out of order occasionally) as well as process can exit before all events are actually sent.

- Ported process classes from RTVS.

### CODE REVIEW

**Comment by AlexanderSher** [comment]
This whole thing looks like a huge overkill to me. In RTVS, we had this method to execute commands in docker on all platforms: https://github.com/Microsoft/RTVS/blob/master/src/Containers/Impl/Docker/LocalDockerService.cs#L174
I don't see much difference between this one and executing python script.

File: src/Core/Impl/OS/ProcessServices.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Python.Core.OS {
+    public sealed class ProcessServices : IProcessServices {
```
Comment (AlexanderSher) [inline src/Core/Impl/OS/ProcessServices.cs]: Do we need to add this service into tests?

Comment (MikhailArkhipov) [reply src/Core/Impl/OS/ProcessServices.cs]: Yes I added it

File: src/Analysis/Core/Impl/Interpreter/PythonLibraryPath.cs [hunk 2]
```diff
@@ -122,59 +122,53 @@ public override string ToString()
...
-                        } catch (OperationCanceledException) {
-                            proc.Kill();
-                            exitCode = -1;
-                        }
-                        if (exitCode != 0) {
-                            throw new InvalidOperationException(string.Join(Environment.NewLine, errorLines));
-                        }
+                var lines = await ps.ExecuteAndCaptureOutputAsync(startInfo, cancellationToken);
+                return lines.Select(s => {
+                    if (s.StartsWithOrdinal(tempWorkingDir, ignoreCase: true)) {
```
Comment (AlexanderSher) [inline src/Analysis/Core/Impl/Interpreter/PythonLibraryPath.cs]: `s.PathStartsWith`

File: src/Analysis/Ast/Impl/Modules/CompiledPythonModule.cs [hunk 3]
```diff
@@ -73,34 +75,29 @@ public override string Documentation
...
+                RedirectStandardOutput = true,
+                RedirectStandardError = true
+            };
+            var ps = Services.GetService<IProcessServices>();
+
+            Log?.Log(TraceEventType.Verbose, "Scrape", startInfo.FileName, startInfo.Arguments);
+            Task<IReadOnlyList<string>> task = null;
+
+            try {
+                task = ps.ExecuteAndCaptureOutputAsync(startInfo);
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Modules/CompiledPythonModule.cs]: IMO, it is better to pass timeout as parameter to `ExecuteAndCaptureOutputAsync`:
```cs
try {
    var token = new CancellationTokenSource(60000).Token;
    var lines = ps.ExecuteAndCaptureOutputAsync(startInfo, token).GetAwaiter().GetResult();
    return string.Join(Environment.NewLine, lines);
} catch (Exception) {
    return string.Empty;
}
```

Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Modules/CompiledPythonModule.cs]: I can create cancellation token with timeout instead

Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Modules/CompiledPythonModule.cs]: Done

**Pull request merged**